lib/srv/db/opensearch/engine.go

```diff
@@ -0,0 +1,287 @@
+	if e.clientConn == nil || err == nil || utils.IsOKNetworkError(err) {
+		return
+	}
+
+	cause := ErrorResponse{
+		Error: ErrorDetails{
+			Reason: err.Error(),
+			Type:   "internal_server_error_exception",
+		},
+		Status: http.StatusInternalServerError,
+	}
+
+	// Different error for access denied case.
+	if trace.IsAccessDenied(err) {
+		cause.Status = http.StatusUnauthorized
+		cause.Error.Type = "access_denied_exception"
+	}
+
+	jsonBody, err := json.Marshal(cause)
+	if err != nil {
+		e.Log.WithError(err).Error("failed to marshal error response")
+		return
+	}
+
+	response := &http.Response{
+		ProtoMajor: 1,
+		ProtoMinor: 1,
+		StatusCode: cause.Status,
+		Body:       io.NopCloser(bytes.NewBuffer(jsonBody)),
+		Header: map[string][]string{
+			"Content-Type":   {"application/json"},
+			"Content-Length": {strconv.Itoa(len(jsonBody))},
+		},
+	}
+
+	if err := response.Write(e.clientConn); err != nil {
+		e.Log.Errorf("OpenSearch error: %+v", trace.Unwrap(err))
+		return
+	}
+}
+
+// HandleConnection authorizes the incoming client connection, connects to the
+// target OpenSearch server and starts proxying requests between client/server.
+func (e *Engine) HandleConnection(ctx context.Context, _ *common.Session) error {
+	err := e.checkAccess(ctx)
+	e.Audit.OnSessionStart(e.Context, e.sessionCtx, err)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+	defer e.Audit.OnSessionEnd(e.Context, e.sessionCtx)
+
+	clientConnReader := bufio.NewReader(e.clientConn)
+	for {
+		req, err := http.ReadRequest(clientConnReader)
+		if err != nil {
+			return trace.Wrap(err)
+		}
+
+		if err := e.process(ctx, req); err != nil {
+			return trace.Wrap(err)
+		}
+	}
+}
+
+// process reads request from connected OpenSearch client, processes the requests/responses and send data back
+// to the client.
+func (e *Engine) process(ctx context.Context, req *http.Request) error {
+	reqCopy, _, err := utils.CloneRequest(req)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	// force HTTPS, set host URL.
+	reqCopy.URL.Scheme = "https"
+	reqCopy.URL.Host = e.sessionCtx.Database.GetURI()
+
+	e.emitAuditEvent(reqCopy)
+
+	roleArn, err := libaws.BuildRoleARN(e.sessionCtx.DatabaseUser, e.sessionCtx.Database.GetAWS().Region, e.sessionCtx.Database.GetAWS().AccountID)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	signCtx := &libaws.SigningCtx{
+		SigningName:   opensearchservice.EndpointsID,
+		SigningRegion: e.sessionCtx.Database.GetAWS().Region,
+		Expiry:        e.sessionCtx.Identity.Expires,
+		SessionName:   e.sessionCtx.Identity.Username,
+		AWSRoleArn:    roleArn,
+		AWSExternalID: e.sessionCtx.Database.GetAWS().ExternalID,
+	}
+
+	signedReq, err := e.signingSvc.SignRequest(e.Context, reqCopy, signCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	tlsConfig, err := e.Auth.GetTLSConfig(ctx, e.sessionCtx)
+	if err != nil {
+		return trace.Wrap(err)
+	}
+
+	clt := &http.Client{
+		Transport: &http.Transport{
+			TLSClientConfig: tlsConfig,
+		},
+	}
```

Comment:
may want to use `lib/defaults.Transport` here and overwrite the returned transport's TLSClientConfig

Reply:
`lib/defaults.Transport` is preferable indeed, but using it has the side-effect of enabling HTTP2. Typically, this is a good thing, but our code didn't support that use case; we copied the server response as-is, including the mismatched protocol version (we only support HTTP1.1 for client requests).

Disabling HTTP2 would work, but instead, I opted to repackage the response:
https://github.com/gravitational/teleport/blob/a6abd933647c01f311ccfafe8214d3005a250b03/lib/srv/db/opensearch/engine.go#L289-L297

Ideally, we would support HTTP2 throughout the stack, but that needs additional work, which ideally would be shared across the various engines that we have, hence the comment:

https://github.com/gravitational/teleport/blob/a6abd933647c01f311ccfafe8214d3005a250b03/lib/srv/db/opensearch/engine.go#L152-L154

Reply:
good point about HTTP2! I should change dynamodb engine to do the same, since it currently uses the default transport and writes the response obtained as-is.